src/Features/CSharp/Portable/UseRecursivePatterns/CSharpAnalyzedNodeRewriter.cs

```diff
@@ -0,0 +1,397 @@
+        {
+            expression = null;
+            return node switch
+            {
+                OrSequence p when CanSimplifyConsecutiveConstantTests(p, out var constants) => AsPatternSyntax(SimplifyConsecutiveConstantTests(constants)),
+                OrSequence p => p.Nodes.Select(p => AsPatternSyntax(p)).Aggregate((left, right) => BinaryPattern(SyntaxKind.OrPattern, left, right)),
+                True _ => DiscardPattern(),
+                False _ => UnaryPattern(DiscardPattern()),
+                Not { Operand: NotNull _ } => ConstantPattern(LiteralExpression(SyntaxKind.NullLiteralExpression)),
+                Not p => UnaryPattern(AsPatternSyntax(p.Operand)),
+                NotNull _ => UnaryPattern(ConstantPattern(LiteralExpression(SyntaxKind.NullLiteralExpression))),
+                Relational p => RelationalPattern(Token(AsSyntaxKind(p.OperatorKind)), (ExpressionSyntax)p.Value.Syntax),
+                Constant p => ConstantPattern(((ExpressionSyntax)p.Value.Syntax).WithoutTrailingTrivia()),
+                Type p => TypePattern(p.TypeSymbol.GenerateTypeSyntax()),
+                Variable p => VarPattern(SingleVariableDesignation(Identifier(p.DeclaredSymbol.Name))),
+                var p when !recurse => AsRecursivePatternSyntax(p, out expression),
+                var p => throw UnexpectedNode(p)
+            };
+        }
+
+        private static Exception UnexpectedNode(AnalyzedNode p)
+        {
+#if DEBUG
+            return new InvalidOperationException($"Unexpected node of type '{p.GetType().Name}'\n{p.Dump()}");
+#else
+            return ExceptionUtilities.UnexpectedValue(p);
+#endif
+        }
+
+        private static SyntaxKind AsSyntaxKind(BinaryOperatorKind kind)
+        {
+            return kind switch
+            {
+                BinaryOperatorKind.LessThan => SyntaxKind.LessThanToken,
+                BinaryOperatorKind.GreaterThan => SyntaxKind.GreaterThanToken,
+                BinaryOperatorKind.LessThanOrEqual => SyntaxKind.LessThanEqualsToken,
+                BinaryOperatorKind.GreaterThanOrEqual => SyntaxKind.GreaterThanEqualsToken,
+                var v => throw ExceptionUtilities.UnexpectedValue(v)
+            };
+        }
+
+        private static PatternSyntax AsRecursivePatternSyntax(AnalyzedNode node, out ExpressionSyntax? expression)
+        {
+            Debug.Assert(!(node is OrSequence), "!(node is OrSequence)");
+
+            ITypeSymbol? type = null;
+            VariableDesignationSyntax? designation = null;
+
+            using var _0 = ArrayBuilder<PositionalPatternClauseSyntax>.GetInstance(out var positionalClauses);
+            using var _1 = ArrayBuilder<SubpatternSyntax>.GetInstance(out var propertySubpatterns);
+            using var _2 = ArrayBuilder<PatternSyntax?>.GetInstance(out var tupleSubpatterns);
+            using var _3 = ArrayBuilder<PatternSyntax?>.GetInstance(out var indexedSubpatterns);
```

Comment:
ITuple matching is not that common but this will make it real easy to add support for list patterns later :)